File: apps/observe-tester/modules/crash-tester/CrashTester.swift
// Copyright 2025-present 650 Industries. All rights reserved.

internal import ExpoModulesCore
import Foundation

/// Crash kinds accepted by `triggerCrash`. Raw values match the TypeScript `CrashKind` union.
enum CrashKind: String, Enumerable {
  /// EXC_BAD_ACCESS / SIGSEGV — dereference of a bogus pointer.
  case badAccess
  /// EXC_CRASH / SIGABRT — Swift `fatalError`.
  case fatalError
  /// EXC_ARITHMETIC / SIGFPE — integer divide by zero.
  case divideByZero
  /// EXC_BAD_INSTRUCTION — force-unwrap of a nil optional.
  case forceUnwrapNil
  /// EXC_BAD_INSTRUCTION — out-of-bounds Swift array access.
  case arrayOutOfBounds
  /// Uncaught Objective-C `NSException`, populates MetricKit's `exceptionReason`.
  case objcException
  /// Stack overflow via unbounded recursion.
  case stackOverflow
}

/// Inline module that exposes crash triggers to the observe-tester app. Test-only:
/// it intentionally crashes the process to exercise the crash-reporting pipeline.
class CrashTester: Module {
  public func definition() -> ModuleDefinition {
    Function("triggerCrash") { (kind: CrashKind) in
      switch kind {
      case .badAccess: CrashTriggers.badAccess()
      case .fatalError: CrashTriggers.fatalErrorCrash()
      case .divideByZero: CrashTriggers.divideByZero()
      case .forceUnwrapNil: CrashTriggers.forceUnwrapNil()
      case .arrayOutOfBounds: CrashTriggers.arrayOutOfBounds()
      case .objcException: CrashTriggers.objcException()
      case .stackOverflow: CrashTriggers.stackOverflow()
      }
    }
  }
}

/// Helpers that intentionally crash the app to produce real MetricKit crash diagnostics.
/// Use only for testing the crash-reporting pipeline.
enum CrashTriggers {
  /// EXC_BAD_ACCESS / SIGSEGV — dereference a bogus pointer.
  static func badAccess() -> Never {
    let pointer = UnsafePointer<Int>(bitPattern: 0x1)!
    _ = pointer.pointee
    fatalError("unreachable")
  }

  /// EXC_CRASH / SIGABRT — Swift fatalError.
  static func fatalErrorCrash() -> Never {
    fatalError("Intentional fatalError for crash-reporting test")
  }

  /// EXC_ARITHMETIC / SIGFPE — integer divide by zero.
  static func divideByZero() -> Never {
    let zero = Int("0")!
    _ = 1 / zero
    fatalError("unreachable")
  }

  /// EXC_BAD_INSTRUCTION — force-unwrap nil.
  static func forceUnwrapNil() -> Never {
    let value: Int? = nil
    _ = value!
    fatalError("unreachable")
  }

  /// Out-of-bounds array access — also EXC_BAD_INSTRUCTION via Swift's runtime trap.
  static func arrayOutOfBounds() -> Never {
    let array: [Int] = []
    _ = array[5]
    fatalError("unreachable")
  }

  /// Uncaught Objective-C exception — produces `exceptionReason` in MetricKit.
  static func objcException() -> Never {
    NSException(
      name: .invalidArgumentException,
      reason: "Intentional NSException for crash-reporting test",
      userInfo: nil
    ).raise()
    fatalError("unreachable")
  }

  /// Stack overflow via unbounded recursion.
  static func stackOverflow() -> Never {
    Self.stackOverflow()
  }
}

File: apps/observe-tester/modules/crash-tester/CrashTester.kt
package expo.modules.crashtester

import android.os.Handler
import android.os.Looper
import android.os.Process
import android.system.Os
import android.system.OsConstants
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.types.Enumerable

/**
 * Crash kinds accepted by `triggerCrash`. Raw values match the TypeScript
 * `CrashKind` union (named after the iOS triggers); each maps to the closest
 * Android equivalent. The `rawValue: String` parameter must stay the only
 * primary-constructor argument — it's what Expo's `Enumerable` converter
 * inspects.
 */
enum class CrashKind(val rawValue: String) : Enumerable {
  /**
   * On Android: the process kills itself with SIGSEGV — a *signaled* death,
   * not a tombstoned native crash. Bare signals are outside the standalone
   * crash allowlist, so this kind tests process-death behavior but produces
   * no stored crash report.
   */
  BAD_ACCESS("badAccess"),
  FATAL_ERROR("fatalError"),
  DIVIDE_BY_ZERO("divideByZero"),
  FORCE_UNWRAP_NIL("forceUnwrapNil"),
  ARRAY_OUT_OF_BOUNDS("arrayOutOfBounds"),

  /** No ObjC on Android — throws a generic runtime exception instead. */
  OBJC_EXCEPTION("objcException"),
  STACK_OVERFLOW("stackOverflow")
}

/**
 * Inline module that exposes crash triggers to the observe-tester app. Test-only:
 * it intentionally crashes the process to exercise the crash-reporting pipeline.
 */
class CrashTester : Module() {
  override fun definition() = ModuleDefinition {
    Function("triggerCrash") { kind: CrashKind ->
      CrashTriggers.trigger(kind)
    }
  }
}

/**
 * Intentionally crashes the app to exercise the real crash-reporting pipeline.
 * Use only for testing.
 *
 * JVM kinds throw from a `Runnable` posted to the main looper — never from the
 * module function body, where expo-modules-core's exception decorator would
 * catch the throw and turn it into a JS error instead of a crash.
 */
object CrashTriggers {
  fun trigger(kind: CrashKind) {
    if (kind == CrashKind.BAD_ACCESS) {
      // Signal-based death is immediate; no JVM throw involved.
      Os.kill(Process.myPid(), OsConstants.SIGSEGV)
      return
    }
    Handler(Looper.getMainLooper()).post {
      throwFor(kind)
    }
  }

  private fun throwFor(kind: CrashKind): Nothing {
    when (kind) {
      CrashKind.FATAL_ERROR ->
        throw RuntimeException("Intentional crash for crash-reporting test")
      CrashKind.DIVIDE_BY_ZERO -> {
        // Computed denominator so the compiler can't reject a constant 1/0.
        val zero = "0".toInt()
        throw IllegalStateException("unreachable: ${1 / zero}")
      }
      CrashKind.FORCE_UNWRAP_NIL -> {
        val value: String? = null
        throw IllegalStateException("unreachable: ${value!!.length}")
      }
      CrashKind.ARRAY_OUT_OF_BOUNDS ->
        throw IllegalStateException("unreachable: ${emptyList<Int>()[5]}")
      CrashKind.OBJC_EXCEPTION ->
        throw IllegalStateException("Intentional exception for crash-reporting test")
      CrashKind.STACK_OVERFLOW ->
        overflow(0)
      CrashKind.BAD_ACCESS ->
        throw IllegalStateException("badAccess is signal-based and handled in trigger()")
    }
  }

  // Accumulates through the result so the recursion can't be optimized away.
  private fun overflow(depth: Int): Nothing {
    overflow(depth + 1)
  }
}
